### ISSUE TITLE
NET-1543 Fix S2583 FP: List is filled in local function, then checking its Count is > 0

### ISSUE THREAD

**Issue by nalka0** [SonarSource/sonar-dotnet#9671]
### Description

There's a false positive for S2583 with the below snippet

### Repro steps

```cs
public static void S2583()
{
    List<int> list = new();
    foreach (var item in Enumerable.Range(0, 5))
    {
        if (item % 2 == 0)
        {
            LocalFunction(item);
        }

        void LocalFunction(int added)
        {
            list.Add(added);
        }
    }

    if (list.Count > 0) // S2583 false positive
    {
        Console.WriteLine("This code is reachable");
    }
}
```

### Expected behavior

No occurence of S2583

### Actual behavior

S2583 is raised on `if (list.Count > 0)`

### Known workarounds

None

### Related information

* Microsoft Visual Studio Community 2022 - Version 17.11.4
* .NET 6 and 8
* SonarScanner for .NET version 9.0.0
* Windows 11

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1543](https://sonarsource.atlassian.net/browse/NET-1543)

[NET-1543]: https://sonarsource.atlassian.net/browse/NET-1543?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
